fmplayer/player.py

```diff
@@ -121,24 +117,32 @@ def play(self, uri):
 
         logger.info('Play Track: {0}'.format(uri))
 
+        # Always block the playlist watcher before trying to play the track
+        # we can then catch play errors and move on gracefully through the
+        # normal stop procedure
+        logger.debug('Block Watcher - STOP_EVENT cleared')
+        STOP_EVENT.clear()  # Reset STOP_EVENT flag to False
+
         try:
             track = self.session.get_track(uri).load()
             self.session.player.load(track)
             self.session.player.play()
-        except (spotify.error.LibError, ValueError):
-            logger.error('Unable to play: {0}'.uri)
-        else:
-            logger.debug('STOP_EVENT cleared')
-            STOP_EVENT.clear()  # Reset STOP_EVENT flag to False
+        except Exception:  # Catch all cos I don't really know what will go wrong here
+            logger.exception('Unable to play {0} - forcing stop'.uri)
+            self.stop()
 
-    def stop(self):
-        """ Stop the current playing track by stopping, and un loading.
+    def stop(self, session=None):
+        """ Fired when a playing track finishes, ensures the tack is unloaded
+        and the ``STOP_EVENT`` is set to ``True``.
         """
 
-        logger.debug('Track Stop - Unload')
-        self.session.player.unload()
-        logger.debug('STOP_EVENT set')
-        STOP_EVENT.set()  # Unblocks the playlist watcher
+        if session is None:
+            session = self.session
```

Comment:
What is difference between session and self.session? Spotify uses only one user instance so session should be same.